File: lib-segment/src/main/java/websiteschema/mpsegment/graph/DijkstraImpl.java
package websiteschema.mpsegment.graph;

import websiteschema.mpsegment.conf.MPSegmentConfiguration;

import java.util.LinkedList;

public class DijkstraImpl implements IShortestPath {

    private final static int numOfVertexes = MPSegmentConfiguration.SectionSize();
    private IGraph graph;
    private int[] route;
    private DijkstraElement dijk = new DijkstraElement(numOfVertexes);

    public DijkstraImpl() {
        route = new int[numOfVertexes];
        clear();
    }

    private void clear() {
        for (int i = 0; i < numOfVertexes; i++) {
            route[i] = 0;
        }
        dijk.init();
    }

    /**
     * find the shortest route or get it from the results directly.
     *
     * @param end - In case reuse the results
     * @return
     * @throws Exception
     */
    @Override
    public Path getShortestPath(int start, int end) {
        clear();
        findShortestRoute(start, end, dijk);
        return new Path(backTracking(end));
    }

    private LinkedList<Integer> backTracking(int end) {
        int last = route[end];
        if (last > 0) {
            LinkedList<Integer> routes = backTracking(last);
            routes.addLast(end);
            return routes;
        }
        LinkedList<Integer> routes = new LinkedList<Integer>();
        routes.add(end);
        return routes;
    }

    private void findShortestRoute(int location, int dest, DijkstraElement dijk) {
        int[] adjacentVertices = graph.getAdjacentVertices(location);
        if (null == adjacentVertices || adjacentVertices.length == 0)
            return;
        for (int neighbor : adjacentVertices) {
            findAndSaveShorterRoute(location, neighbor, dijk);
        }
        int nextVertex = dijk.findNewShortestPath();
        if (nextVertex >= 0) {
            if (nextVertex != dest) {
                findShortestRoute(nextVertex, dest, dijk);
            }
        }
    }

    private void findAndSaveShorterRoute(final int location, final int vertex, DijkstraElement dijk) {
        int distance = getEdgeWeight(location, vertex) + (dijk.hasFoundShortestPathTo(location) ? dijk.getDistanceOfPathTo(location) : 0);
        int knownDistance = dijk.getDistanceOfPathTo(vertex);
        if (distance < knownDistance) {
            dijk.setDistanceOfPathTo(vertex, distance);
            dijk.reached(vertex);
            route[vertex] = location;
        }
    }

    protected int getEdgeWeight(int location, int vertex) {
        return graph.getEdgeWeight(location, vertex);
    }

    @Override
    public void setGraph(IGraph graph) {
        this.graph = graph;
    }

    protected IGraph getGraph() {
        return graph;
    }

    protected int[] getRouteBackTrace() {
        return route;
    }

    protected DijkstraElement getDijkstraElement() {
        return dijk;
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/graph/DijkstraImpl.scala
package websiteschema.mpsegment.graph

import websiteschema.mpsegment.conf.MPSegmentConfiguration

class DijkstraImpl extends IShortestPath {

  private val numOfVertexes = MPSegmentConfiguration.SectionSize
  private var graph: IGraph = null
  private val route = new Array[Int](numOfVertexes)
  private val dijk = DijkstraElement(numOfVertexes)

  clear()

  private def clear() {
    for (i <- 0 until numOfVertexes) {
      route(i) = 0
    }
    dijk.init()
  }

  /**
   * find the shortest route or get it from the results directly.
   *
   * @param end - In case reuse the results
   * @return
   * @throws Exception
   */
  override def getShortestPath(start: Int, end: Int): Path = {
    clear()
    findShortestRoute(start, end, dijk)
    return Path(backTracking(end))
  }

  private def backTracking(end: Int): List[Int] = {
    val last = route(end)
    if (last > 0) {
      val routes = backTracking(last)
      return routes ::: List[Int](end)
    }
    return List[Int](end)
  }

  private def findShortestRoute(location: Int, dest: Int, dijk: DijkstraElement) {
    val adjacentVertices = graph.getAdjacentVertices(location)
    if (null == adjacentVertices || adjacentVertices.length == 0)
      return
    for (neighbor <- adjacentVertices) {
      findAndSaveShorterRoute(location, neighbor, dijk)
    }
    val nextVertex = dijk.findNewShortestPath()
    if (nextVertex >= 0) {
      if (nextVertex != dest) {
        findShortestRoute(nextVertex, dest, dijk)
      }
    }
  }

  private def findAndSaveShorterRoute(location: Int, vertex: Int, dijk: DijkstraElement) {
    val distance = getEdgeWeight(location, vertex) + (if(dijk.hasFoundShortestPathTo(location)) dijk.getDistanceOfPathTo(location) else 0)
    val knownDistance = dijk.getDistanceOfPathTo(vertex)
    if (distance < knownDistance) {
      dijk.setDistanceOfPathTo(vertex, distance)
      dijk.reached(vertex)
      route(vertex) = location
    }
  }

  def getEdgeWeight(location: Int, vertex: Int): Int = {
    return graph.getEdgeWeight(location, vertex)
  }

  override def setGraph(graph: IGraph) {
    this.graph = graph
  }

  def getGraph(): IGraph = {
    return graph
  }

  def getRouteBackTrace(): Array[Int] = {
    return route
  }

  def getDijkstraElement(): DijkstraElement = {
    return dijk
  }
}
